_posts/2016/2016-09-08-kubernetes-preparing-nodes.md

```diff
@@ -0,0 +1,259 @@
+title: "Kubernetes HA setup part 2: Preparing the nodes"
+date: "2016-09-08 07:16:17"
+image: blog-header/kubernetes-docker-flanneld-logo.jpg
+categories: tech-stories
+authors: ["Karsten P."]
+---
+
+The second part of the blog post series deals with the preparation of the Kubernetes and Service nodes.
+It describes the installation of the HAProxy, the Flannel daemon (flanneld), the docker engine (docker) and kubelet. 
+
+## Install the Service node
+
+The Service node which is located outside the cluster undertakes the task of a load balancer.
+Since the open source version of nginx does not support TCP loadbalancing we use [HAProxy](http://www.haproxy.org/).
+HAProxy distributes packages between the hosts on the [transport layer](https://en.wikipedia.org/wiki/Transport_layer) of the [OSI reference model](https://en.wikipedia.org/wiki/OSI_model).
+The following configuration is set up after the [Installation of HAProxy](https://haproxy.debian.net/).
+
+{% highlight bash %}
+root@service:~$ cat > /etc/haproxy/haproxy.cfg << EOF
+global
+    log         127.0.0.1 local2
+
+    chroot      /var/lib/haproxy
+    pidfile     /var/run/haproxy.pid
+    maxconn     4000
+    user        haproxy
+    group       haproxy
+    daemon
+
+    # turn on stats unix socket
+    stats socket /var/lib/haproxy/stats
+
+defaults
+        log     global
+        mode    http
+        option  dontlognull
+        option  redispatch
+        retries 3
+        maxconn 2000
+        timeout connect         10s
+        timeout client          1m
+        timeout server          1m
+
+listen https :443
+        mode tcp
+        balance roundrobin
+
+        server apiserver01 master01.example.com:443 check
+        server apiserver01 master02.example.com:443 check
+EOF
+root@service:~$ systemd restart haproxy
+{% endhighlight %}
+
+After restarting the service it should be possible to reach the master nodes via the load balancer URL (e.g. *loadbalancer.example.com*) although at this time there is no service available on port 443.
+
+## Flanneld
+
+[Flannel](https://coreos.com/flannel) enables the cluster wide distribution of the pods' packages.
+Since the description of the networks are stored in *etcd* a connection of all Kubernetes nodes (master and worker) to the *etcd* cluster has to be enabled.
+The docker daemon on each Kubernetes node will be performed with special parameters provided by flanneld.
+Cross-node communication of containers is realized by iptables that route the packages to the correct target servers.
+For this purpose flanneld creates its own virtual network interface and must therefore run with root rights.
+
+### Create *etcd* client certificates
+
+Kubernetes nodes need a client certificate because each user of the *etcd* cluster has to authenticate.
+It can be created like described in the first part of this blog post series but there are differences in the _openssl.cnf_.
+It is a bit simpler since flanneld has to show itself as client only.
```

Comment:
The _openssl.cnf_ has a simpler structure as flanneld has to identify itself as client only.